armory/scenarios/poisoning_witches_brew.py

```diff
@@ -0,0 +1,260 @@
+            raise ValueError(
+                "source_class should have one element or be the same length as trigger_index"
+            )
+
+        for i, trigger_ind in enumerate(self.trigger_index):
+            if y_test[trigger_ind] != self.source_class[i]:
+                raise ValueError(
+                    f"Trigger image {i} does not belong to source class (class {y_test[trigger_ind]} != class {self.source_class[i]})"
+                )
+
+        if self.use_poison:
+
+            attack_config["kwargs"]["percent_poison"] = adhoc_config[
+                "fraction_poisoned"
+            ]
+            attack_config["kwargs"]["source_class"] = self.source_class
+            attack_config["kwargs"]["target_class"] = self.target_class
+
+            data_filepath = (
+                attack_config["kwargs"].pop("data_filepath")
+                if "data_filepath" in attack_config["kwargs"].keys()
+                else None
+            )
+
+            attack = config_loading.load_attack(attack_config, self.model)
+            if data_filepath is not None:
+                data_filepath = os.path.join(
+                    paths.runtime_paths().dataset_dir, data_filepath
+                )
+            self.poisoner = DatasetPoisonerWitchesBrew(
+                attack,
+                x_test,
+                y_test,
+                self.source_class,
+                self.target_class,
+                self.trigger_index,
+                data_filepath,
+            )
+            self.test_poisoner = self.poisoner
+
+    def load_metrics(self):
+        self.non_trigger_accuracy_metric = metrics.MetricList("categorical_accuracy")
+        self.trigger_accuracy_metric = metrics.MetricList("categorical_accuracy")
+
+        self.benign_test_accuracy_per_class = (
+            {}
+        )  # store accuracy results for each class
+        if self.config["adhoc"].get("compute_fairness_metrics", False):
+            self.fairness_metrics = FairnessMetrics(
+                self.config["adhoc"], self.use_filtering_defense, self
+            )
+        else:
+            log.warning(
+                "Not computing fairness metrics.  If these are desired, set 'compute_fairness_metrics':true under the 'adhoc' section of the config"
+            )
+
+    def load_dataset(self, eval_split_default="test"):
+        # Over-ridden because we need batch_size = 1 for the test set for this attack.
+
+        dataset_config = self.config["dataset"]
+        dataset_config = copy.deepcopy(dataset_config)
+        dataset_config["batch_size"] = 1
+        eval_split = dataset_config.get("eval_split", eval_split_default)
+        log.info(f"Loading test dataset {dataset_config['name']}...")
+        self.test_dataset = config_loading.load_dataset(
+            dataset_config,
+            split=eval_split,
+            num_batches=self.num_eval_batches,
+            **self.dataset_kwargs,
+        )
+        self.i = -1
+
```

Comment:
Move this before `load_metrics` so that it is in the same order as `poison.py`.

Reply:
Done